### PR TITLE
Add the initial implementation of volumetric lighting

### PR DESCRIPTION
**Pull request by EvgeniiG**

### CODE REVIEW

File: ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Common.hlsl [hunk 1]
```diff
@@ -499,6 +504,20 @@ float LinearEyeDepth(float3 positionWS, float4x4 viewProjMatrix)
     return mul(viewProjMatrix, float4(positionWS, 1.0)).w;
 }
 
+// 'z' is the view-space Z position (linear depth).
+// saturate() the output of the function to clamp them to the [0, 1] range.
+float EncodeLogarithmicDepth(float z, float4 encodingParams)
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Common.hlsl]: add comment about what is in encodingParams

File: ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Filtering.hlsl [hunk 2]
```diff
@@ -0,0 +1,44 @@
+#ifndef UNITY_FILTERING_INCLUDED
+#define UNITY_FILTERING_INCLUDED
+
+#if !defined(SHADER_API_GLES)
+
+// Cardinal (interpolating) B-Spline of the 2nd degree (3rd order). Support = 3x3.
+// The fractional coordinate of each part is assumed to be in the [0, 1] range (centered on 0.5).
+// https://www.desmos.com/calculator/47j9r9lolm
+float2 BSpline2IntLeft(float2 x)
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Filtering.hlsl]: use real in this file

File: ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/GeometricTools.hlsl [hunk 3]
```diff
@@ -88,6 +88,91 @@ float3 RayPlaneIntersect(in float3 rayOrigin, in float3 rayDirection, in float3
     return rayOrigin + rayDirection * dist;
 }
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/GeometricTools.hlsl]: use real

File: ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/GeometricTools.hlsl [hunk 4]
```diff
@@ -88,6 +88,91 @@ float3 RayPlaneIntersect(in float3 rayOrigin, in float3 rayDirection, in float3
...
+    float q = -0.5 * (b + FastSign(b) * sqrt(d));
+    roots   = float2(q / a, c / q);
+
+    return (d >= 0);
+}
+
+// Can support cones with an elliptic base: pre-scale 'coneAxisX' and 'coneAxisY' by (h/r_x) and (h/r_y).
+// Returns parametric distances 'tEntr' and 'tExit' along the ray,
+// subject to constraints 'tMin' and 'tMax'.
+bool IntersectRayCone(float3 rayOrigin,  float3 rayDirection,
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/GeometricTools.hlsl]: let's try to be consistent with naming, other function in this file are BoxRayIntersect, SphereRayIntersect, either rename yours or their

File: ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/GeometricTools.hlsl [hunk 5]
```diff
@@ -88,6 +88,91 @@ float3 RayPlaneIntersect(in float3 rayOrigin, in float3 rayDirection, in float3
...
+
+    // Check for grazing intersections.
+    if (tEntr == tExit) { hit = false; }
+
+    return hit;
+}
+
+// This implementation does not attempt to explicitly handle NaNs.
+// Ref: https://tavianator.com/fast-branchless-raybounding-box-intersections-part-2-nans/
+bool IntersectRayAABB(float3 rayOrigin, float3 rayDirection,
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/GeometricTools.hlsl]: we have already this fast implementation, try to rely on it: BoxRayIntersect

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/Volumetrics/HomogeneousFog.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
+namespace UnityEngine.Experimental.Rendering.HDPipeline
+{
+    [ExecuteInEditMode]
+    [AddComponentMenu("Rendering/Homogeneous Fog", -1)]
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/Volumetrics/HomogeneousFog.cs]: GameObject/RenderPipeline/High Definition/Homogenous Fog

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/Volumetrics/VolumetricLighting.cs [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+        properties.scattering = GetScatteringCoefficient();
+        properties.extinction = GetExtinctionCoefficient();
+
+        return properties;
+    }
+} // class VolumeParameters
+
+public partial class HDRenderPipeline : RenderPipeline
+{
+    public enum VolumetricLightingPreset
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/Volumetrics/VolumetricLighting.cs]: Don't create a partial class, Volumetric should be a module like the lightloop

Comment (sebastienlagarde) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/Volumetrics/VolumetricLighting.cs]: we can discuss about this,  suppose you have done partial class for falicity ?

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/Volumetrics/VolumetricLighting.cs [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+
+        // Swap with the last element and shrink the array.
+        m_ViewIdArray[viewOffset] = m_ViewIdArray[lastOffset];
+        m_ViewCount--;
+    }
+
+    // Uses a logarithmic depth encoding.
+    // Near plane: depth = 0; far plane: depth = 1.
+    // x = n, y = log2(f/n), z = 1/n, w = 1/log2(f/n).
+    public static Vector4 ComputeLogarithmicDepthEncodingParams(float nearPlane, float farPlane)
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/Volumetrics/VolumetricLighting.cs]: Move to core ? as it is use with a common.hlsl funciotn

File: ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/GeometricTools.hlsl (hunk 3 above)
Comment (EvgeniiG) [reply ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/GeometricTools.hlsl]: We don't appear to use real inside GeometricTools.hlsl.

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/Volumetrics/Resources/VolumetricLighting.compute [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+CBUFFER_START(UnityVolumetricLighting)
+    float4   _VBufferSampleOffset;              // {x, y, z}, w = rendered frame count
+    float4x4 _VBufferCoordToViewDirWS;          // Actually just 3x3, but Unity can only set 4x4
+CBUFFER_END
+
+//--------------------------------------------------------------------------------------------------
+// Implementation
+//--------------------------------------------------------------------------------------------------
+
+struct Ray
```
Comment (sebastienlagarde) [inline ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/Volumetrics/Resources/VolumetricLighting.compute]: Move ray to geometricTools.hlsl with GetPointAtDistance

File: ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Common.hlsl (hunk 1 above)
Comment (EvgeniiG) [reply ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Common.hlsl]: Done

File: ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Filtering.hlsl (hunk 2 above)
Comment (EvgeniiG) [reply ScriptableRenderPipeline/Core/CoreRP/ShaderLibrary/Filtering.hlsl]: Done

File: ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/Volumetrics/HomogeneousFog.cs (hunk 6 above)
Comment (EvgeniiG) [reply ScriptableRenderPipeline/HDRenderPipeline/HDRP/Lighting/Volumetrics/HomogeneousFog.cs]: Moved to "RenderPipeline/High Definition/Homogenous Fog"

**Pull request merged**